### ISSUE TITLE
Classes that are not disposable should be guarded

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/center-and-variability#463]
For code review #447 ...

When I set up the checklist in https://github.com/phetsims/center-and-variability/issues/447#issuecomment-1683022865, I noticed that this item was added after the self code review #449.  It was added as a result of the work in Disposable, and for the Kepler’s Laws code review.

> - [ ] All classes that are not disposable should either (a) use `isDisposable: false`, or (b) implement a `dispose` method that calls `Disposable.assertNotDisposable`.  Use (a) for classes that inherit a `dispose` method. Use (b) for classes that do not inherit a `dispose` method.

I see no occurrences of `isDisposable` or `Disposable.assertNotDisposable` in center-and-variability or soccer-common. So unless `dispose` is supported by all classes (which would be very unusually for sim-specific code) CaV is not compliant.

**Comment by samreid** [comment]
I do not believe that code review requirement to be complete. There should be an option (c) that says "or (c) do not have a `dispose` method at all." For example, `AsyncCounter.ts` doesn't extend PhetioObject or have a `dispose` method on its interface. If anyone tries to call `dispose` on it, it would be a type error. Can we update the code review checklist to allow option (c)?

**Comment by pixelzoom** [comment]
I agree. But I don't think we need (c), I think we need to delete (b). I don't see the point in adding a method like:

```ts
public dispose(): void {
  Disposable.assertNotDisposable();
}
```

I've updated the title of this issue, and I'll modify the code-review template.

**Comment by pixelzoom** [comment]
Actually, we do need (b).  Here's the revised checklist item:

> - [ ] All classes that do not properly implement `dispose` should either (a) use `isDisposable: false`, or (b) implement a `dispose` method that calls `Disposable.assertNotDisposable`.  Use (a) for classes that inherit a `dispose` method from Disposable. Use (b) for classes that inherit a `dispose` method from something other than Disposable.  The goal here is to prevent clients from calling `dispose` for something that does not properly clean itself up, creating a memory leak.  This is especially important for common code, but is good defensive programming for sim-specific code.

**Comment by pixelzoom** [comment]
One more comment about:
> This is especially important for common code, but is good defensive programming for sim-specific code.

I think it's esssential to address this in soccer-common, which will be used in other sims, by other developers.

I think it's also valuable and worth the time for future maintenance of center-and-variability. But if you want to exercise "developer discretion", it's up to you.

**Comment by samreid** [comment]
In soccer common, I added `isDisposable: false` to every site that has a `dispose()` method on the interface. Other objects cannot be disposed, so do not need anything additional.

In center-and-variability, I'm concerned that adding `isDisposable: false` and `override dispose(){assertion error}` everywhere there is a `dispose` interface would be unfortunate and unnecessary (over 60 sites). Instead, I added a lint rule that prevents the word "dispose" from being mentioned in the Center and Variability *.ts files. That way, if a center-and-variability maintainer tries to dispose something, they will get a lint error right away, before even running the code. The main situation where this would be problematic is if center-and-variability becomes a common code repo for another sim that wants to deal with dispose differently. However, there is no plan on the horizon for center-and-variability to become a common code repo for other sims.  So I feel this is a reasonable solution.  I'll check in with @marlitas and @matthew-blackman before we bring this back to @pixelzoom.

**Comment by pixelzoom** [comment]
I recommend adding `isDisposable: false` where it's appropriate in center-and-variability, rather than doing something entirely different for this sim.  It will take you less time to add to the 60 sites that it will to continue discussing, or to negotiate the pros/cons of an alternative lint rule.  

The work on Disposable was done for a purpose, and we had a lot of discussions about it.  I'd be interested to hear how @zepumph feels about this, since he did most of the heavy lifting for Disposable.

**Comment by pixelzoom** [comment]
soccer-common looks great.

I recommend adding `isDisposable: false` where it's appropriate in center-and-variability, rather than doing something entirely different for this sim.  It will take you less time to add to the 60 sites that it will to continue discussing, or to negotiate the pros/cons of an alternative lint rule.  

In general, if you create a  subclass, and do something in the subclass such that an inherited method is broken (incorrect, incomplete,...), the proper thing to do is to handle that in your code, typically by overriding the method.

So... PhET developers should get in the habit of using `isDisposable`. This feature of Disposable is a nice convenience that avoids the need to override.  A lot of thought/work went into it, and we had many discussions about it, a developer PSA, etc.  I'd be interested to hear how @zepumph feels about this, since he did most of the heavy lifting for Disposable.

**Comment by samreid** [comment]
I added `isDisposable: false` throughout center-and-variability. I did not repeat `isDisposable: false` in subclasses where it is already marked in a superclass.

**Comment by marlitas** [comment]
We decided to keep the lint rule on top of the `isDisposable: false` implementation. We all agreed we would like to keep the lint rule as well. Over to @pixelzoom for review.

**Comment by pixelzoom** [comment]
👍🏻
